### ISSUE TITLE
Duplicate diagnostics in unpacking targets

### ISSUE THREAD

**Issue by dhruvmanila** [astral-sh/ty#185]
### Summary

Consider the following code:
```py
class Foo():
    pass

a, b = Foo().nonexistant
```

Here, the RHS of the assignment has an error `unresolved-attribute` which is being shown 3 times (2 + 1 where 2 is the number of variables involved in the unpacking):

<img width="1043" alt="Image" src="https://github.com/user-attachments/assets/e6be58cc-cc5e-4079-814c-de25191bc10d" />

This is because the standalone expression is being inferred _outside_ the unpacking in `infer_assignment_definition` (and similarly in `infer_for_statement_definition`). This inference should only be done for non-unpack targets.

Fixing this would lead to another bug which is that the diagnostics raised during the unpacking query isn't being merged into the outer type inference builder.

**Comment by dhruvmanila** [comment]
This should be relatively easy to fix once https://github.com/astral-sh/ruff/pull/18041 has landed. This is because that PR adds a `Definition` ingredient for all attributes which would then allow us to use the `infer_unpack_types` for all attributes involved in an unpacking.

**Comment by dhruvmanila** [comment]
I tried fixing this today based on the assumption that assignments containing an attribute expressions now always create a `Definition` but that isn't true. It's only created if `PlaceExpr::try_from` is successful, so a call expression like the following doesn't create a `Definition`. I'll have to re-think the approach that I used here.

```py
class Foo:
    def __init__(self):
        self.x = 1

Foo().x = 2
```
